pkg/natdiscovery/request_handle.go

```diff
@@ -0,0 +1,133 @@
+	}
+
+	if req.Receiver.GetEndpointId() != nd.localEndpoint.Spec.CableName {
+		klog.Warningf("Received NAT discovery packet for endpoint %q, but we are endpoint %q "+
+			"if the port for NAT discovery has been mapped somewhere an error may exist", req.Receiver.GetEndpointId(),
+			nd.localEndpoint.Spec.CableName)
+
+		response.Response = proto.ResponseType_UNKNOWN_DST_ENDPOINT
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if !nd.isKnownEndpoint(req.Sender.GetEndpointId(), req.Sender.GetClusterId()) {
+		// This can be ok for some time, because one cluster could discover an endpoint earlier on the broker than the
+		// the other endpoint has discovered him
+		klog.V(log.DEBUG).Infof("Received a NAT discovery packet from an endpoint %q on cluster %q we don't know about yet",
+			req.Receiver.GetEndpointId(), req.Receiver.GetClusterId())
+
+		response.Response = proto.ResponseType_UNKNOWN_SRC_ENDPOINT
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if req.UsingSrc.GetIP() != "" && req.UsingSrc.GetIP() != addr.IP.String() {
+		klog.V(log.DEBUG).Infof("Received NAT packet from endpoint %q, cluster %q, where NAT has been detected, "+
+			"source IP changed",
+			req.Sender.GetEndpointId(), req.Sender.GetClusterId())
+		klog.V(log.DEBUG).Infof("Original src IP was %q, received src IP is %q", req.UsingSrc.IP, addr.IP.String())
+
+		response.Response = proto.ResponseType_SRC_MODIFIED
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if int(req.UsingSrc.Port) != addr.Port {
+		klog.V(log.DEBUG).Infof("Received NAT packet from endpoint %q, cluster %q, where NAT on the source has been detected, "+
+			"src port changed",
+			req.Sender.GetEndpointId(), req.Sender.GetClusterId())
+		klog.V(log.DEBUG).Infof("Original src IP was %q, received src IP is %q", req.UsingSrc.IP, addr.IP.String())
+
+		response.Response = proto.ResponseType_SRC_MODIFIED
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	response.Response = proto.ResponseType_OK
+
+	return nd.sendResponseToAddress(&response, addr)
+}
+
+func (nd *natDiscovery) isKnownEndpoint(endpointID, clusterID string) bool {
+	nd.Lock()
+	defer nd.Unlock()
+
+	ep, ok := nd.remoteEndpoints[endpointID]
+
+	return ok && ep.Endpoint().Spec.ClusterID == clusterID
+}
+
+func (nd *natDiscovery) sendResponseToAddress(response *proto.SubmarinerNatDiscoveryResponse, addr *net.UDPAddr) error {
+	msgResponse := proto.SubmarinerNatDiscoveryMessage_Response{Response: response}
+	message := proto.SubmarinerNatDiscoveryMessage{Message: &msgResponse}
+	buf, err := proto2.Marshal(&message)
+	if err != nil {
+		return errors.Wrapf(err, "Error marshaling response %#v", response)
+	}
+
+	if length, err := nd.serverUDPWrite(buf, addr); err != nil {
+		return errors.Wrapf(err, "Error sending response packet %#v", response)
+	} else if length != len(buf) {
+		return errors.Errorf("The sent UDP packet was smaller than requested, sent=%d, expected=%d", length, len(buf))
```

Comment:
```suggestion
		return errors.Errorf("the sent UDP packet was smaller than requested, sent=%d, expected=%d", length, len(buf))
```

Reply:
ok so @tpantelis explained to me, that in wrap we should use low-case, and on the base Errors, upcase.

Reply:
Ack. I'm fine either way. I just remember seeing shipyard/e2e/framework/util.go using lower case for Errors.

Reply:
> Ack. I'm fine either way. I just remember seeing shipyard/e2e/framework/util.go using lower case for Errors.

yeah it's recommended [here](https://github.com/golang/go/wiki/CodeReviewComments#error-strings). We have a linter that will at least flag it with `fmt.Errorf` but it doesn't seem to recognize "_github.com/pkg/errors_".  When logging a message via klog then you capitalize.

Reply:
done